--- a/lib/accelerated/x86/aes-padlock.c
+++ b/lib/accelerated/x86/aes-padlock.c
@@ -62,7 +62,7 @@
 	struct padlock_ctx *ctx = _ctx;
 	struct padlock_cipher_data *pce;
 #ifdef HAVE_LIBNETTLE
-	struct aes_ctx nc;
+	struct aes256_ctx nc;
 #endif
 
 	memset(_ctx, 0, sizeof(struct padlock_cipher_data));
@@ -89,12 +89,12 @@
 	      common_24_32:
 		/* expand key using nettle */
 		if (ctx->enc)
-			aes_set_encrypt_key(&nc, keysize, userkey);
+			aes256_set_encrypt_key(&nc, userkey);
 		else
-			aes_set_decrypt_key(&nc, keysize, userkey);
+			aes256_set_decrypt_key(&nc, userkey);
 
 		memcpy(pce->ks.rd_key, nc.keys, sizeof(nc.keys));
-		pce->ks.rounds = nc.rounds;
+		pce->ks.rounds = _AES256_ROUNDS;
 
 		pce->cword.b.keygen = 1;
 		break;
